### PR TITLE
ARROW-5681: [FlightRPC] Add Flight-specific error APIs

### PR DESCRIPTION
**Pull request by ghost**
This adds a set of exceptions (in Java/Python) and status codes (in C++) for Flight, to convey Flight-specific issues and to let server implementations provide more semantic error messages to clients.

The error codes defined are a subset of the gRPC ones. In Flight/Java, you could always raise `io.grpc.StatusRuntimeException` yourself, but we want to abstract away from the gRPC APIs. 

There are not tests for conveying these statuses cross-language. That should be done as part of ARROW-5875.

Travis: https://travis-ci.com/lihalite/arrow/builds/118452703
AppVeyor: https://ci.appveyor.com/project/lihalite/arrow/builds/25849820

### CODE REVIEW

File: cpp/src/arrow/flight/internal.cc
Comment (emkornfield) [inline cpp/src/arrow/flight/internal.cc]: I believe I added an already exists error code.

File: cpp/src/arrow/flight/internal.cc
Comment (emkornfield) [inline cpp/src/arrow/flight/internal.cc]: most if not all statuses can be constructed directly with:
Status::IOError("message ...", variable, "message ...").with_detail(details)

File: cpp/src/arrow/flight/internal.cc
Comment (emkornfield) [inline cpp/src/arrow/flight/internal.cc]: Unknown might be better?

File: cpp/src/arrow/flight/internal.cc
Comment (emkornfield) [inline cpp/src/arrow/flight/internal.cc]: are you avoiding using Status.ToString() because it included Arrow's interpretation?

File: cpp/src/arrow/flight/internal.cc
Comment (lihalite) [reply cpp/src/arrow/flight/internal.cc]: Thanks, that's much cleaner.

File: cpp/src/arrow/flight/internal.cc
Comment (lihalite) [reply cpp/src/arrow/flight/internal.cc]: Agreed, changed. I also added a missing branch for gRPC's UNKNOWN.

File: cpp/src/arrow/flight/internal.cc
Comment (lihalite) [reply cpp/src/arrow/flight/internal.cc]: Yes, otherwise the error gets a bit repetitive.

File: cpp/src/arrow/flight/internal.cc
Comment (lihalite) [reply cpp/src/arrow/flight/internal.cc]: Thanks for the pointer, I've changed that.

Do you think NOT_FOUND should also be promoted to the general status codes, as the counterpart?

Comment (emkornfield) [reply cpp/src/arrow/flight/internal.cc]: I think KeyError might be the equivalent.

Comment (lihalite) [reply cpp/src/arrow/flight/internal.cc]: Ah I see, thanks - updated.

**Comment by lihalite** [comment]
Rebased/fixed conflicts.

File: cpp/src/arrow/flight/internal.cc
Comment (emkornfield) [inline cpp/src/arrow/flight/internal.cc]: FAILED_PRECONDITION, ABORTED, and RESOURCE_EXHAUSTED  errors might be good candidates to make top-level.  We should probably discuss on the mailing list.  But they have slightly different semantics and might be useful to build more general purpose machinery around handling them

**Comment by emkornfield** [comment]
@lihalite this is kind of a big change.  If no one gets to it sooner I can try to have a more thorough review done sometime over the weekend but I have other competing priorities at the moment.

**Comment by lihalite** [comment]
@emkornfield no worries, thanks for the comments so far. I'm not in a particular rush - over the next week is fine! I am trying to land this (+the follow up proposal on the mailing list) for 1.0.0, since it is a little painful to not really be able to tell what kind of error happened without inspecting repr/str.

File: python/pyarrow/_flight.pyx
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: I _think_ all the raises should be under the `with gil`, though perhaps Cython does it for us?

File: cpp/src/arrow/flight/types.h
Comment (pitrou) [inline cpp/src/arrow/flight/types.h]: Could simply return a `shared_ptr<FlightStatusDetail>` and use `nullptr` to signal that unwrapping failed.

**Review by pitrou — COMMENTED** [review]
Very nice!

File: python/pyarrow/_flight.pyx
Comment (lihalite) [reply python/pyarrow/_flight.pyx]: It "worked" before but I'd rather be explicit, thanks.

File: cpp/src/arrow/flight/types.h
Comment (lihalite) [reply cpp/src/arrow/flight/types.h]: Good point, thanks. I've gotten too used to making everything an out parameter by default...

**Comment by lihalite** [comment]
Rebased, Travis build: https://travis-ci.com/lihalite/arrow/builds/119628842

File: python/pyarrow/_flight.pyx
Comment (pitrou) [inline python/pyarrow/_flight.pyx]: Should we inherit from ArrowException here rather than below?

Comment (lihalite) [reply python/pyarrow/_flight.pyx]: I had to do this dance to make Cython happy: ArrowException isn't cdef, so inheriting from it from a cdef class doesn't work.

Comment (pitrou) [reply python/pyarrow/_flight.pyx]: Ok, that's fine with me.

**Review by pitrou — APPROVED** [review]
+1. Are you looking for a Java review?

**Comment by lihalite** [comment]
@pitrou thanks! Yes, still waiting on a Java review.

File: java/flight/src/main/java/org/apache/arrow/flight/grpc/StatusUtils.java
Comment (emkornfield) [inline java/flight/src/main/java/org/apache/arrow/flight/grpc/StatusUtils.java]: per comment above these still seem like INVALID_ARGUMENT might be too broad a catch-all.

File: java/flight/src/test/java/org/apache/arrow/flight/auth/TestBasicAuth.java
Comment (emkornfield) [inline java/flight/src/test/java/org/apache/arrow/flight/auth/TestBasicAuth.java]: it might pay to making a "testing" utility class that combines these to lines together.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightStatusCode.java
Comment (emkornfield) [inline java/flight/src/main/java/org/apache/arrow/flight/FlightStatusCode.java]: please document what the null parameter represent.

File: java/flight/src/main/java/org/apache/arrow/flight/CallStatus.java
Comment (emkornfield) [inline java/flight/src/main/java/org/apache/arrow/flight/CallStatus.java]: can you expand on this comment.  In particular, it is not immediately obvious why this can't just be an exception class (i.e. why do we need to represent OK)?  It seems like this might have do with the flight state machine if I understand the code correctly.

File: java/flight/src/main/java/org/apache/arrow/flight/CallStatus.java
Comment (emkornfield) [inline java/flight/src/main/java/org/apache/arrow/flight/CallStatus.java]: consider making a construct that just take code, so you don't have to make nullable parameters for constructing from the the enum.

**Comment by emkornfield** [comment]
Overall, seems ok to me on the java side, left a few comment.  I'm not entirely sure everything should be translated to a runtime exception but we can revisit this if it becomes an issue later.  I still think it might be discussing exposing a few more Status codes at the top-level instead of conflating them, but this is also something we can evolve as we more usage on flight (hopefully we get  feedback before we declare flight stable otherwise the migration might be a little painful).

File: java/flight/src/main/java/org/apache/arrow/flight/CallStatus.java
Comment (lihalite) [reply java/flight/src/main/java/org/apache/arrow/flight/CallStatus.java]: I've mostly followed the gRPC API in this regard, which has both checked and unchecked versions of the exception that use a common status object. We don't need to represent OK and don't use the status, so I've removed it. I did expand the comment a bit.

File: java/flight/src/main/java/org/apache/arrow/flight/grpc/StatusUtils.java
Comment (lihalite) [reply java/flight/src/main/java/org/apache/arrow/flight/grpc/StatusUtils.java]: I added these two codes to the Flight codes, to mirror what was done in C++/Python.

File: java/flight/src/test/java/org/apache/arrow/flight/auth/TestBasicAuth.java
Comment (lihalite) [reply java/flight/src/test/java/org/apache/arrow/flight/auth/TestBasicAuth.java]: Good idea, thanks.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightStatusCode.java
Comment (lihalite) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightStatusCode.java]: I ended up creating the single-parameter constructor as suggested below.

**Comment by lihalite** [comment]
Thanks for the review! I addressed your feedback and rebased. The Java API is modeled off of the gRPC API, but that isn't necessarily a requirement (e.g we could use checked exceptions if we feel that makes more sense).

**Review by emkornfield — APPROVED** [review]
LGTM.  Waiting for travis to be green before merging.

**Comment by lihalite** [comment]
Rebased again, personal Travis: https://travis-ci.com/lihalite/arrow/builds/120745269

**Comment by pitrou** [comment]
Thanks @lihalite !

**Comment by lihalite** [comment]
Thanks all for looking over this! On to the next task :)

**Pull request closed without merging**
